### PR TITLE
fix(spooler): Proactively populate the project cache for envelopes spooled to disk

### PR DESCRIPTION
**Pull request by olksdr**
When the Relay restarts the index of the spooled to disk envelopes must be restored:
* read all the project keys we spooled to disk
* generate unique index
* make sure the project cache fetches them

Once the project state received Relay will start to unspool saved to disk envelopes. 

This should allow Relay to unspool faster, even for the projects which do not have the high volume traffic. 

related to https://github.com/getsentry/team-ingest/issues/257

#skip-changelog

### CODE REVIEW

File: relay-server/src/actors/spooler/mod.rs [hunk 1]
```diff
@@ -684,6 +697,47 @@ impl OnDisk {
...
+
+        // Fold the list into the index format.
+        let index = extracted_keys?.into_iter().fold(
+            BTreeMap::new(),
+            |mut acc: BTreeMap<ProjectKey, BTreeSet<QueueKey>>, key| {
+                acc.entry(key.own_key).or_default().insert(key);
+                acc.entry(key.sampling_key).or_default().insert(key);
+                acc
+            },
+        );
```
Comment (olksdr) [inline relay-server/src/actors/spooler/mod.rs]: I could do this in one go, without allocating a vec, but that means I will have to also ignore the errors here, and filter them out, and report them in `extract_key` instead.

File: relay-server/src/actors/spooler/mod.rs [hunk 2]
```diff
@@ -1028,13 +1091,34 @@ impl BufferService {
         Ok(())
     }
 
+    /// Handles the retrieving the index from the underlying spool.
+    async fn handle_get_index(&mut self, _: GetIndex) -> Result<(), BufferError> {
```
Comment (iker-barriocanal) [inline relay-server/src/actors/spooler/mod.rs]: Do you think what the "handling" is, is relevant enough to document it?

File: relay-server/src/actors/spooler/mod.rs [hunk 3]
```diff
@@ -684,6 +697,55 @@ impl OnDisk {
...
+                None
+            }
+        }
+    }
+
+    /// Returns the index from the on-disk spool.
+    async fn get_spooled_index(
+        &self,
+    ) -> Result<BTreeMap<ProjectKey, BTreeSet<QueueKey>>, BufferError> {
+        let keys = sql::get_keys()
```
Comment (iker-barriocanal) [inline relay-server/src/actors/spooler/mod.rs]: Some time ago we discussed how querying the whole database during startup could be problematic for Relay. Is this no longer the case? Are there new considerations we've made?

**Review by iker-barriocanal — COMMENTED** [review]
The code looks good to me. Leaving one question about the performance I think we should address before moving forward with this.

File: relay-server/src/actors/spooler/mod.rs (hunk 2 above)
Comment (olksdr) [reply relay-server/src/actors/spooler/mod.rs]: Added some more docs. Let me know what exactly you would like to see here.

File: relay-server/src/actors/spooler/mod.rs (hunk 3 above)
Comment (olksdr) [reply relay-server/src/actors/spooler/mod.rs]: I've changed this a little bit. We do not block the start up , and execute the query in the background, meaning we allow ourself to run it a bit longer and process the result async.

**Review by iker-barriocanal — APPROVED** [review]
Awesome! Thank you for the nice PR 😄

**Review by Dav1dde — APPROVED** [review]

File: relay-server/src/services/project_cache.rs [hunk 4]
```diff
@@ -996,6 +1021,9 @@ impl Service for ProjectCacheService {
                 }
             };
 
+            // Reqest the existing index from the spooler.
```
Comment (jjbayer) [inline relay-server/src/services/project_cache.rs]:
```suggestion
            // Request the existing index from the spooler.
```

File: relay-server/src/services/spooler/mod.rs [hunk 5]
```diff
@@ -1029,13 +1092,49 @@ impl BufferService {
...
+                    match OnDisk::get_spooled_index(&db).await {
+                        Ok(index) => {
+                            relay_log::trace!(
+                                "recover index from disk with {} unique project keys",
+                                index.len()
+                            );
+                            project_cache.send(SpoolIndex(index))
+                        }
+                        Err(err) => {
+                            relay_log::error!("failed to retrive the index from the disk: {err}")
```
Comment (jjbayer) [inline relay-server/src/services/spooler/mod.rs]:
```suggestion
                            relay_log::error!("failed to retrieve the index from the disk: {err}")
```

File: relay-server/src/services/spooler/mod.rs [hunk 6]
```diff
@@ -183,6 +183,10 @@ impl RemoveMany {
 #[derive(Debug)]
 pub struct Health(pub Sender<bool>);
 
+/// Requests the index [`ProjectKey`] -> [`QueueKey`] of the data currently residing in the spool.
+#[derive(Debug)]
+pub struct GetIndex;
```
Comment (jjbayer) [inline relay-server/src/services/spooler/mod.rs]: nit: Is there a more descriptive name we can use for this message. Something like `TriggerProjectStateFetchForSpooledEnvelopes`, but simpler?

**Review by jjbayer — COMMENTED** [review]
Is there a way we can test this behavior?

File: relay-server/src/services/spooler/mod.rs (hunk 6 above)
Comment (Dav1dde) [reply relay-server/src/services/spooler/mod.rs]: ~~I like `GetIndex`, it just gives you the entire spool index and describes that operation, the project cache then just uses the index.~~

Oh actually I thought you would pass a return channel with the `GetIndex` but it actually just calls into the project cache directly. Agree then a more descriptive name wouldn't hurt.

Comment (olksdr) [reply relay-server/src/services/spooler/mod.rs]: how about now, see changes in commit [`710e80f`](https://github.com/getsentry/relay/pull/2962/commits/710e80ffcb1b95e13b7b6ef1f8a9f6a9ebb0364a)

Added more docs and also rename the messages to hopefully give more intuition about what's going on

File: relay-server/src/services/spooler/mod.rs [hunk 7]
```diff
@@ -1029,13 +1098,49 @@ impl BufferService {
...
+    /// Handles retrieving of the index from the underlying spool.
+    ///
+    /// To process this request we spawn the tokio task and once it's done the result will be sent
+    /// to [`ProjectCache`].
+    ///
+    /// If the spool is memory based, we ignore this request - the index in the
+    /// [`ProjectCache`] should be full and correct.
+    /// If the spool is located on the disk, we read up the keys and compile the index of the
+    /// spooled envelopes for [`ProjectCache`].
+    async fn handle_get_index(&mut self, _: RestoreIndex) -> Result<(), BufferError> {
```
Comment (jjbayer) [inline relay-server/src/services/spooler/mod.rs]: Should we rename this to `handle_restore_index` as well?

**Review by jjbayer — APPROVED** [review]
> how about now, see changes in commit [710e80f](https://github.com/getsentry/relay/pull/2962/commits/710e80ffcb1b95e13b7b6ef1f8a9f6a9ebb0364a)

"Restore" sounds good to me!

File: relay-server/src/services/spooler/mod.rs (hunk 7 above)
Comment (olksdr) [reply relay-server/src/services/spooler/mod.rs]: yep, will do it now

**Pull request merged**